okhttp/src/main/java/okhttp3/internal/ws/WebSocketOptions.kt

```diff
@@ -0,0 +1,158 @@
+
+    private val COMPRESSION_NO_TAKEOVER = WebSocketOptions(
+      compressionEnabled = true,
+      contextTakeover = false
+    )
+
+    private val COMPRESSION_WITH_TAKEOVER = WebSocketOptions(
+      compressionEnabled = true,
+      contextTakeover = true
+    )
+
+    @Throws(IOException::class)
+    @JvmStatic
+    fun parseServerResponse(response: Response): WebSocketOptions {
+      // No extension header - server does not support permessage-deflate.
+      val header = response.header(HEADER_WEB_SOCKET_EXTENSION) ?: return NO_COMPRESSION
+
+      // Server is free to return empty header to indicate no compression
+      // See end of https://tools.ietf.org/html/rfc7692#section-5 chapter
+      if (header.isEmpty()) {
+        return NO_COMPRESSION
+      }
+
+      val extension = header
+        .split(", ")
+        // Additional or fallback extensions are not supported currently.
+        // See https://tools.ietf.org/html/rfc7692#section-5.2 for details.
+        .getOrNull(0)
+
+      if (extension.isNullOrBlank()) {
+        throw ProtocolException("$HEADER_WEB_SOCKET_EXTENSION malformed: '$header'")
+      }
+
+      val tokens = extension.split("; ").map { it.replace(";", "") }
+      if (tokens.isEmpty()) {
+        throw ProtocolException("Extension could not be parsed: '$extension'")
+      }
+
+      val extensionName = tokens[0]
+      if (extensionName != EXTENSION_PERMESSAGE_DEFLATE) {
+        // Client MUST fail for extension it did not ask for
+        throw ProtocolException("Extension not supported: '$extensionName'")
+      }
+
+      val options = tokens.drop(1).map(Option.Companion::parse)
+      if (options.isEmpty()) {
+        // Server did not force client_no_context_takeover,
+        // so client can use compression with context takeover.
+        return COMPRESSION_WITH_TAKEOVER
+      }
+
+      if (options.toSet().size != options.size) {
+        // Client MUST fail for duplicate options.
+        throw ProtocolException("Duplicate options found in '$extension'")
+      }
+
+      val clientNoContextTakeover = Option.CLIENT_NO_CONTEXT_TAKEOVER in options
+
+      return if (clientNoContextTakeover) {
+        COMPRESSION_NO_TAKEOVER
+      } else {
+        COMPRESSION_WITH_TAKEOVER
+      }
+    }
+  }
+}
+
+private enum class Option(val id: String) {
+  CLIENT_NO_CONTEXT_TAKEOVER("client_no_context_takeover"),
+  SERVER_NO_CONTEXT_TAKEOVER("server_no_context_takeover"),
+  CLIENT_MAX_WINDOW_BITS("client_max_window_bits"),
+  SERVER_MAX_WINDOW_BITS("server_max_window_bits");
+
+  companion object {
+    /**
+     * If server decides to use MAX_WINDOW_BITS less than 15, is fine.
+     * The deflater has enough information from the deflated data
+     * to parse it correctly.
+     */
+    private const val MIN_MAX_WINDOW_BITS = 8
+    private const val MAX_MAX_WINDOW_BITS = 15
+
+    private const val SUPPORTED_CLIENT_MAX_WINDOW_BITS = MAX_MAX_WINDOW_BITS
+
+    fun parse(option: String): Option =
```

Comment:
I think we’ll be happier if we never offer the `client_max_window_bits` or `server_max_window_bits` parameters, which makes them forbidden in the agreement response.